src/main/java/org/folio/service/PasswordResetLinkServiceImpl.java

```diff
@@ -83,90 +83,104 @@ public PasswordResetLinkServiceImpl(ConfigurationClient configurationClient, Aut
     this.userPasswordService = userPasswordService;
   }
 
-  @Override
-  public Future<String> sendPasswordRestLink(String userId, OkapiConnectionParams connectionParams) {
+  public Future<String> sendPasswordRestLink(User user, Map<String, String> okapiHeaders) {
+    OkapiConnectionParams connectionParams = new OkapiConnectionParams(okapiHeaders);
     Holder<Map<String, String>> configMapHolder = new Holder<>();
-    Holder<User> userHolder = new Holder<>();
     Holder<String> passwordResetActionIdHolder = new Holder<>();
     Holder<Boolean> passwordExistsHolder = new Holder<>();
     Holder<String> linkHolder = new Holder<>();
 
     return configurationClient.lookupConfigByModuleName(MODULE_NAME, GENERATE_LINK_REQUIRED_CONFIGURATION, connectionParams)
       .compose(configurations -> {
         configMapHolder.value = configurations;
-        return userModuleClient.lookupUserById(userId, connectionParams);
+        if (StringUtils.isBlank(user.getUsername())) {
+          String message = "User without username cannot reset password";
+          UnprocessableEntityMessage entityMessage = new UnprocessableEntityMessage("user.absent-username", message);
+          throw new UnprocessableEntityException(Collections.singletonList(entityMessage));
+        }
+        return isPasswordExists(user.getId(), connectionParams, configMapHolder, passwordResetActionIdHolder);
       })
+      .compose(passwordExists -> signToken(connectionParams, passwordExists, passwordExistsHolder, passwordResetActionIdHolder))
+      .compose(token -> sendNotification(connectionParams, token, configMapHolder, linkHolder, passwordExistsHolder, user))
+      .map(v -> linkHolder.value);
+
+  }
+
+  @Override
+  public Future<String> sendPasswordRestLink(String userId, Map<String, String> okapiHeaders) {
+    OkapiConnectionParams connectionParams = new OkapiConnectionParams(okapiHeaders);
+    return userModuleClient.lookupUserById(userId, connectionParams)
       .compose(optionalUser -> {
-        if (!optionalUser.isPresent()) {
+        if (optionalUser.isEmpty()) {
           String message = String.format("User with id '%s' not found", userId);
           UnprocessableEntityMessage entityMessage = new UnprocessableEntityMessage("user.not-found", message);
           throw new UnprocessableEntityException(Collections.singletonList(entityMessage));
```

Comment:
It's better to wrap exceptions in the FailedFuture to process it in further chain blocks

Reply:
Refactored